tokio-util/src/sync/mpsc.rs

```diff
@@ -0,0 +1,145 @@
+pub struct Sender<T: 'static> {
+    // field order: `state` can contain reference to `inner`,
+    // so it must be dropped first.
+    /// State of this sender
+    state: State<T>,
+    /// Underlying channel
+    inner: tokio::sync::mpsc::Sender<T>,
+    /// Pinned marker
+    pinned: PhantomPinned,
+}
+
+impl<T: 'static> Sender<T> {
+    /// It is OK to get mutable reference to state, because it is not
+    /// self-referencing.
+    fn pin_project_state(self: Pin<&mut Self>) -> &mut State<T> {
+        unsafe { &mut Pin::into_inner_unchecked(self).state }
+    }
+
+    /// Sender must be pinned because state can contain references to it
+    fn pin_project_inner(self: Pin<&mut Self>) -> Pin<&mut tokio::sync::mpsc::Sender<T>> {
+        unsafe { self.map_unchecked_mut(|t| &mut t.inner) }
+    }
+}
+
+impl<T: 'static> Clone for Sender<T> {
+    fn clone(&self) -> Self {
+        Sender {
+            state: State::Empty,
+            inner: self.inner.clone(),
+            pinned: PhantomPinned,
+        }
+    }
+}
+
+type AcquireFutOutput<T> = Result<Permit<'static, T>, SendError<()>>;
+
+enum State<T: 'static> {
+    /// We do not have permit, but we didn't start acquiring it.
+    Empty,
+    /// We have a permit to send.
+    // ALERT: this is self-reference to the Sender.
+    Ready(Permit<'static, T>),
+    /// We are in process of acquiring a permit
+    // ALERT: contained future contains self-reference to the sender.
+    Acquire(Pin<Box<dyn Future<Output = AcquireFutOutput<T>>>>),
+}
+
+impl<T: 'static> std::fmt::Debug for State<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            State::Empty => f.debug_tuple("Empty").finish(),
+            State::Ready(_) => f.debug_tuple("Ready").field(&"_").finish(),
+            State::Acquire(_) => f.debug_tuple("Acquire").field(&"_").finish(),
+        }
+    }
+}
+
+impl<T: 'static> Sender<T> {
+    /// Wraps a [tokio sender](tokio::sync::mpsc::Sender).
+    pub fn new(inner: tokio::sync::mpsc::Sender<T>) -> Self {
+        Sender {
+            inner,
+            state: State::Empty,
+            pinned: PhantomPinned,
+        }
+    }
+
+    /// Returns sender readiness state
+    pub fn is_ready(&self) -> bool {
+        matches!(self.state, State::Ready(_))
+    }
+
+    /// Sends a message.
+    /// This function can only be called when `is_ready() == false`,
+    /// otherwise it panics.
+    pub fn send(self: Pin<&mut Self>, value: T) {
+        let permit = match std::mem::replace(self.pin_project_state(), State::Empty) {
+            State::Ready(permit) => permit,
+            _ => panic!("called send() on non-ready Sender"),
+        };
+        permit.send(value);
+    }
+
+    /// Disarms permit, allowing other senders to use freed capacity slot.
+    /// This function can only be called when `is_ready() == true`,
+    /// otherwise it panics.
+    pub fn disarm(mut self: Pin<&mut Self>) {
```

Comment:
Should this panic if not ready? The 0.2 sender had the following comment:

> Returns false if no slot is reserved for this sender (usually because poll_ready was not previously called, or did not succeed).

Reply:
done